frontend/src/AnnotatedText.tsx

```diff
@@ -0,0 +1,60 @@
+import React from 'react';
+import * as firebase from "firebase/app";
+import {MedicalAnnotation, AnnotationEntity} from "./MedicalAnnotation"
+
+
+interface AnnotationResult {
+    data: {
+        Entities: Array<AnnotationEntity>
+    }
+}
+
+interface AnnotationProps {
+    message: string
+}
+
+interface AnnotationCache {
+    [indexer: string] : AnnotationResult
+}
+const cache: AnnotationCache = {}
+
+export default function AnnotatedText(props: AnnotationProps)  {
+    const [result, setResult] = React.useState<AnnotationResult | null>(null);
+
+    React.useEffect(() => {
+        let annotateTranscription = firebase.functions().httpsCallable('annotateTranscription');
+        if (props.message in cache) {
+            console.log("Cached!");
+            setResult(cache[props.message]);
+        } else {
+            annotateTranscription({'message': props.message})
+                .then((response: AnnotationResult) => {
+                    cache[props.message] = response;
+                    setResult(response);
+                });
```

Comment:
I would write this as:

    const getAnnotation = (() => {
          const cache = {} as AnnotationCache;
          return async (message: string) => {
              const result = cache[message] || await annotateTranscription({ message });
              cache[message] = result;
              return result;
        };
    })();

    export default function AnnotatedText({ message }: AnnotationProps)  {
    ...
      useEffect(() => {
        getAnnotation(message).then(setResult);
      }, [message]);

That is:

- deconstruct the argument to the component function
- move the caching mechanism outside the component 
- close the cache inside the caching mechanism
- use `await` rather than `then` (in the caching mechanism — async effects aren't allowed for some stupid reason)
- import `useEffect` and `useState`.
- prefer "as" to constant-type declaration
- simplify the caching mechanism
- tell React to only invoke the effect when `message` changes

Reply:
Done.